### PR TITLE
Refactor difficulty icons to not suck

### PR DESCRIPTION
**Pull request by peppy**
Something I've been putting off for a while. The hierarchy of these icons and the very sparse and different usages made it a nightmare to deal with.

- `DifficultyIcon` is now a very simple implementation with a bindable to update the rating externally, and some sane defaults.
- `CalculatingDifficultyIcon` is responsible for doing automatic calculation when on-screen.
- `DifficultyRetriever` is a stand-alone component that handles the retrieval. I figure this may be useful in other contexts in the future. Or not. At least it's well defined and documented now.
- I was able to delete/combine some classes along the way (`GroupedFilterable` and `Filterable` combined, and some weird `MatchModeType` nuked).

### CODE REVIEW

File: osu.Game/Beatmaps/Drawables/DifficultyRetriever.cs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        /// </summary>
+        /// <param name="beatmapInfo">The beatmap to use for calculation.</param>
+        /// <param name="ruleset">The ruleset to use for calculation.</param>
+        /// <param name="mods">The mods to use for calculation.</param>
+        public DifficultyRetriever(IBeatmapInfo beatmapInfo, IRulesetInfo ruleset, IReadOnlyList<Mod> mods)
+        {
+            this.beatmapInfo = beatmapInfo;
+            this.ruleset = ruleset;
+            this.mods = mods;
+        }
```
Comment (peppy) [inline osu.Game/Beatmaps/Drawables/DifficultyRetriever.cs:59]: I'm not sure if this is worth keeping around. It is not used anywhere right now. It is also the only usage of:

https://github.com/ppy/osu/blob/f8830c6850128456266c82de83273204f8b74ac0/osu.Game/Beatmaps/BeatmapDifficultyCache.cs#L104-L117

File: osu.Game/Overlays/BeatmapSet/BeatmapPicker.cs [hunk 2]
```diff
@@ -274,8 +274,9 @@ public DifficultySelectorButton(APIBeatmap beatmapInfo)
                             Alpha = 0.5f
                         }
                     },
-                    icon = new DifficultyIcon(beatmapInfo, shouldShowTooltip: false)
+                    icon = new DifficultyIcon(beatmapInfo.Ruleset)
                     {
+                        Current = { Value = new StarDifficulty(beatmapInfo.StarRating, 0) },
```
Comment (frenzibyte) [inline osu.Game/Overlays/BeatmapSet/BeatmapPicker.cs:279]: Is it intended that this doesn't just use `DifficultyIcon(BeatmapInfo)` and disable tooltip explicitly?

File: osu.Game/Beatmaps/Drawables/DifficultyIcon.cs [hunk 3]
```diff
@@ -124,69 +111,23 @@ private void load(OsuColour colours)
...
 
             if (onlineID >= 0 && rulesets.GetRuleset(onlineID.Value)?.CreateInstance() is Ruleset rulesetInstance)
                 return rulesetInstance.CreateIcon();
 
             return new SpriteIcon { Icon = FontAwesome.Regular.QuestionCircle };
         }
 
-        ITooltip<DifficultyIconTooltipContent> IHasCustomTooltip<DifficultyIconTooltipContent>.GetCustomTooltip() => new DifficultyIconTooltip();
+        ITooltip<DifficultyIconTooltipContent> IHasCustomTooltip<DifficultyIconTooltipContent>.
+            GetCustomTooltip() => new DifficultyIconTooltip();
```
Comment (frenzibyte) [inline osu.Game/Beatmaps/Drawables/DifficultyIcon.cs:128]: Was this automated by rider? I've also seen this happening locally on any file I can touch recently, not sure if something changed in the configuration somewhere.

**Review by frenzibyte — COMMENTED** [review]
~~Upon testing, it seems difficulties that are beyond the song select screen area but not completely offscreen will remain unpopulated, making it look off:~~

https://user-images.githubusercontent.com/22781491/175451666-01086c42-2a0d-4139-a6ad-d50ac31cb735.mp4

~~(notice the difficulty cards at the bottom right of the screen updating SR the moment they enter song select range)~~

---

Never mind, I thought this PR introduced the usage of `DelayedLoadUnloadWrapper` for processing difficulty, while it's actually already present in master.

File: osu.Game/Beatmaps/Drawables/DifficultyRetriever.cs (hunk 1 above)
Comment (frenzibyte) [reply osu.Game/Beatmaps/Drawables/DifficultyRetriever.cs:59]: I think it'll be useful for cases like queued up `DrawableRoomPlaylistItem` which are not currently selected but need to display the correct SR icon in the item.

File: osu.Game/Overlays/BeatmapSet/BeatmapPicker.cs (hunk 2 above)
Comment (peppy) [reply osu.Game/Overlays/BeatmapSet/BeatmapPicker.cs:279]: This was before I added the `ShowTooltip`. I'll update accordingly.

File: osu.Game/Beatmaps/Drawables/DifficultyIcon.cs (hunk 3 above)
Comment (peppy) [reply osu.Game/Beatmaps/Drawables/DifficultyIcon.cs:128]: I'd say so.

File: osu.Game/Beatmaps/Drawables/DifficultyRetriever.cs (hunk 1 above)
Comment (peppy) [reply osu.Game/Beatmaps/Drawables/DifficultyRetriever.cs]: Is now gone anyway.

**Comment by peppy** [comment]
I've fixed all the reported issues. See https://github.com/ppy/osu/pull/18819/commits/28837693e5a72ccf692df76a7bc231a473987c4b commit message for important dialogue.

**Review by smoogipoo — APPROVED** [review]

**Review by frenzibyte — APPROVED** [review]

**Pull request merged**